### PR TITLE
[12.0][add][stock_picking_move_done_manual]

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
This PR adds 3 modules:

stock_picking_move_done_manual
===========================
This module allows to manually complete individual moves from a picking
for the quantity picked without requiring to validate the whole picking
and create a backorder.

#. Create a new picking and indicate products and quantities.
#. If you want to process all quantities at once, press the button 'Validate'
   on the picking. You will be able to confirm as usual, and the picking
   will be completed.
#. If you want to partially process the picking, set the done quantities you
   wish to process and then press the "Complete" check button on the lines
   that you wish to complete. The move will be completed, and if there were
   quantities pending to transfer, they will be proposed in a separate line.
   If you process all moves manually the transfer will be automatically set
   as completed by default.
#. If you want to complete the transfer with partially moved quantities you
   will be asked to complete it with a backorder or without.

* stock_move_action_done_hook
* stock_picking_button_validate_hook

### CODE REVIEW

File: stock_picking_move_done_manual/readme/CONFIGURE.rst [hunk 1]
```diff
@@ -0,0 +1,14 @@
+To configure this module, you need to:
```
Comment (AdriaGForgeFlow) [inline stock_picking_move_done_manual/readme/CONFIGURE.rst]: I think this does not belong here

File: stock_picking_button_validate_hook/model/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+        def new_button_validate(self):
+            self.ensure_one()
+            if not self.move_lines and not self.move_line_ids:
+                raise UserError(_('Please add some items to move.'))
+            # If no lots when needed, raise error
+            picking_type = self.picking_type_id
+            # START OF PATCH #
+            no_quantities_done = self._button_validate_get_no_quantities_done()
+            # no_quantities_done = all(
```
Comment (AdriaGForgeFlow) [inline stock_picking_button_validate_hook/model/stock_picking.py]: Remove this commented code, already in the function `_button_validate_get_no_quantities_done`

**Comment by AdriaGForgeFlow** [comment]
Currently if a stock move has been done the user cannot Unreserve the quantities reserved for other stock moves. If you are now managing stock moves separately, shouldn't it be possible to unreserve even some stock moves have been done? What do you think?

Peek 2020-04-06 15-56

File: stock_picking_move_done_manual/models/stock_move.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+from odoo import api, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _action_done_get_picking(self, moves, moves_todo):
+        picking = super(StockMove, self)._action_done_get_picking(
+            moves, moves_todo)
+        for move in moves.filtered(lambda m: m.state != 'cancel'):
```
Comment (HviorForgeFlow) [inline stock_picking_move_done_manual/models/stock_move.py:16]: Maybe you could ensure all moves have same picking, like moves.filtered(...).mapped('picking_id'). 
Code works anyway.

**Review by HviorForgeFlow — COMMENTED** [review]
I'm agree with @AdriaGForgeFlow , it seems a hook on method_do_unreserve is required to handle accordingly.

You should drop also stock_picking_move_done_manual/readme/CONFIGURE.rst

File: stock_picking_move_done_manual/readme/CONFIGURE.rst (hunk 1 above)
Comment (JordiBForgeFlow) [reply stock_picking_move_done_manual/readme/CONFIGURE.rst]: @AdriaGForgeFlow Thanks. Comment attended.

File: stock_picking_button_validate_hook/model/stock_picking.py (hunk 2 above)
Comment (JordiBForgeFlow) [reply stock_picking_button_validate_hook/model/stock_picking.py]: @AdriaGForgeFlow Thanks. Comment attended.

**Comment by JordiBForgeFlow** [comment]
@AdriaGForgeFlow nice catch. I have implemented the logic to allow unreserving.

**Comment by AdriaGForgeFlow** [comment]
After doing some stock moves manually if I want to Validate the rest of the picking, Odoo is asking me to create a backorder even if all products are reserved:

backorder

**Comment by rousseldenis** [comment]
@JordiBForgeFlow What's the status of this ?

**Pull request closed without merging**